File: src/sttp/sttp-core/Application/Validation/TreeSitterValidator.cs
using SttpMcp.Domain.Contracts;
using SttpMcp.Domain.Models;

namespace SttpMcp.Application.Validation;

public sealed class TreeSitterValidator : INodeValidator
{
    private static readonly string[] RequiredLayers = ["⊕⟨", "⦿⟨", "◈⟨", "⍉⟨"];
    private static readonly string[] ValidTiers = ["raw", "daily", "weekly", "monthly", "quarterly", "yearly"];

    public ValidationResult Validate(string rawNode)
    {
        if (string.IsNullOrWhiteSpace(rawNode))
            return ValidationResult.Fail("Node is empty", ValidationFailureReason.ParseFailure);

        foreach (var layer in RequiredLayers)
        {
            if (!rawNode.Contains(layer))
                return ValidationResult.Fail($"Missing required layer: {layer}", ValidationFailureReason.MissingLayer);
        }

        var provIdx = rawNode.IndexOf("⊕⟨", StringComparison.Ordinal);
        var envIdx = rawNode.IndexOf("⦿⟨", StringComparison.Ordinal);
        var conIdx = rawNode.IndexOf("◈⟨", StringComparison.Ordinal);
        var metIdx = rawNode.IndexOf("⍉⟨", StringComparison.Ordinal);

        if (!(provIdx < envIdx && envIdx < conIdx && conIdx < metIdx))
            return ValidationResult.Fail(
                "Layer order violation — required: ⊕ → ⦿ → ◈ → ⍉",
                ValidationFailureReason.SchemaViolation);

        var tierValid = ValidTiers.Any(t => rawNode.Contains($"tier: {t}", StringComparison.OrdinalIgnoreCase));
        if (!tierValid)
            return ValidationResult.Fail(
                $"Invalid tier — expected one of: {string.Join("|", ValidTiers)}",
                ValidationFailureReason.InvalidTier);

        var contentBlock = ExtractContentBlock(rawNode);
        if (contentBlock is not null)
        {
            var depth = MaxNestingDepth(contentBlock);
            if (depth > 5)
                return ValidationResult.Fail(
                    $"Content nesting depth {depth} exceeds maximum of 5",
                    ValidationFailureReason.NestingDepth);
        }

        return ValidationResult.Ok();
    }

    public bool VerifyPsi(SttpNode node)
    {
        var computed = node.CompressionAvec?.Psi ?? 0.0f;
        return Math.Abs(computed - node.Psi) < 0.01f;
    }

    private static string? ExtractContentBlock(string raw)
    {
        var start = raw.IndexOf("◈⟨", StringComparison.Ordinal);
        var end = raw.IndexOf("⍉⟨", StringComparison.Ordinal);
        if (start < 0 || end < 0 || end <= start) return null;
        return raw[start..end];
    }

    private static int MaxNestingDepth(string block)
    {
        var max = 0;
        var depth = 0;
        foreach (var c in block)
        {
            if (c == '{')
            {
                depth++;
                max = Math.Max(max, depth);
            }
            else if (c == '}')
            {
                depth--;
            }
        }

        return max;
    }
}

File: src/sttp/sttp-core-rs/src/application/validation/tree_sitter_validator.rs
use crate::domain::contracts::NodeValidator;
use crate::domain::models::{AvecState, SttpNode, ValidationFailureReason, ValidationResult};

const REQUIRED_LAYERS: [&str; 4] = ["⊕⟨", "⦿⟨", "◈⟨", "⍉⟨"];
const VALID_TIERS: [&str; 6] = ["raw", "daily", "weekly", "monthly", "quarterly", "yearly"];

#[derive(Debug, Default, Clone, Copy)]
pub struct TreeSitterValidator;

impl NodeValidator for TreeSitterValidator {
    fn validate(&self, raw_node: &str) -> ValidationResult {
        if raw_node.trim().is_empty() {
            return ValidationResult::fail("Node is empty", ValidationFailureReason::ParseFailure);
        }

        for layer in REQUIRED_LAYERS {
            if !raw_node.contains(layer) {
                return ValidationResult::fail(
                    format!("Missing required layer: {layer}"),
                    ValidationFailureReason::MissingLayer,
                );
            }
        }

        let prov_idx = raw_node.find("⊕⟨").unwrap_or(usize::MAX);
        let env_idx = raw_node.find("⦿⟨").unwrap_or(usize::MAX);
        let con_idx = raw_node.find("◈⟨").unwrap_or(usize::MAX);
        let met_idx = raw_node.find("⍉⟨").unwrap_or(usize::MAX);

        if !(prov_idx < env_idx && env_idx < con_idx && con_idx < met_idx) {
            return ValidationResult::fail(
                "Layer order violation - required: ⊕ -> ⦿ -> ◈ -> ⍉",
                ValidationFailureReason::SchemaViolation,
            );
        }

        let lower = raw_node.to_ascii_lowercase();
        let tier_valid = VALID_TIERS
            .iter()
            .any(|t| lower.contains(&format!("tier: {t}")));
        if !tier_valid {
            return ValidationResult::fail(
                format!("Invalid tier - expected one of: {}", VALID_TIERS.join("|")),
                ValidationFailureReason::InvalidTier,
            );
        }

        if let Some(content_block) = extract_content_block(raw_node) {
            let depth = max_nesting_depth(content_block);
            if depth > 5 {
                return ValidationResult::fail(
                    format!("Content nesting depth {depth} exceeds maximum of 5"),
                    ValidationFailureReason::NestingDepth,
                );
            }
        }

        ValidationResult::ok()
    }

    fn verify_psi(&self, node: &SttpNode) -> bool {
        let computed = node.compression_avec.unwrap_or_else(AvecState::zero).psi();
        (computed - node.psi).abs() < 0.01
    }
}

fn extract_content_block(raw: &str) -> Option<&str> {
    let start = raw.find("◈⟨")?;
    let end = raw.find("⍉⟨")?;
    if end <= start {
        None
    } else {
        Some(&raw[start..end])
    }
}

fn max_nesting_depth(block: &str) -> usize {
    let mut max_depth = 0usize;
    let mut depth = 0usize;

    for c in block.chars() {
        if c == '{' {
            depth += 1;
            max_depth = max_depth.max(depth);
        } else if c == '}' {
            depth = depth.saturating_sub(1);
        }
    }

    max_depth
}
